### ISSUE TITLE
Running tests locally

### ISSUE THREAD

**Issue by bjarkihall** [jimmymcp/al-test-runner#26]
Hi and thanks for a great extension!
I've been wondering for a while if it's possible to run the tests locally instead of in a container?

I know it's docker-based:
https://github.com/jimmymcp/al-test-runner/blob/master/PowerShell/Invoke-RunTests.ps1

which invokes:
https://github.com/microsoft/navcontainerhelper/blob/master/AppHandling/Run-TestsInNavContainer.ps1

which then uses:
https://github.com/microsoft/navcontainerhelper/blob/master/AppHandling/PsTestFunctions.ps1

So I've been wondering if the local option would be possible by optionally running the scripts straight from PsTestFunctions instead of doing it inside a container?
This is generally a question if it's possible and how you would approach it in this repo, would I even be able to submit a PR?

**Comment by jimmymcp** [comment]
Hi Bjarki,

I'm sure it would be possible to run those scripts against a local installation, rather than a Docker container. I don't have time to investigate that myself at the moment, but you'd be very welcome to implement it yourself and submit a PR.

I suppose there would need to a new config key to indicate that the BC server isn't running inside a container. Ideally we'd reuse some functions in the bccontainerhelper function to run the tests but I'm not aware that there are any functions for running tests outside containers?

**Comment by bjarkihall** [comment]
Yeah, PsTestFunctions's "Run-Tests" interacts with a client context that uses the service dlls to interact with the test page.

I was going to reference the file directly but some paths are hardcoded, since BcContainerHelper assumes certain things to be in certain places, but it seems like it would technically be possible to write a file like Run-TestsInNavContainer.ps1 which just runs a script similar to PsTestFunctions.ps1 straight away, instead of using Invoke-ScriptInBCContainer.

I might experiment by cloning this repo and Freddy's scripts and see where it leads me (e.g. how much of the scripts could be left unmodified) and I was mainly curious how these kind of files would fit within the project before starting, and if you'd tried local setup before. Thanks for your feedback. :) I'll let you know if I do this.

I've also heard rumors that MS is going to implement a testing dev-endpoint with an API for vscode, so maybe I'll just be waiting for that. Might not be soon though since there are still things missing like code-coverage, etc.

Btw. do you want me to close the issue or keep it open?

**Comment by jimmymcp** [comment]
Thanks. It might make sense to speak to Freddy about this first, see what he thinks about including a new function in bccontainerhelper to run tests against a local service tier. He might also have more info about when they will implement a new way to execute tests and whether it is worth pursuing this in the meantime.

We can leave this issue open for now to remind me the conversation is ongoing!

**Comment by freddydk** [comment]
What version are you going to run tests in?

**Comment by freddydk** [comment]
Run-Tests has a parameter called -connectFromHost and -useUrl - with those two set, the function is almost able to run without a container (it still copies something from a container I think, but you could find those on the dvd - newtonsoft and one more)

**Comment by bjarkihall** [comment]
> What version are you going to run tests in?

We're currently only planning on testing on latest releases, so we're on BC 17 now and update as MS releases new versions - so there's no need to backport to e.g. versions 14/15/16, if that helps.

**Comment by jimmymcp** [comment]
I haven't tested this but I think the recent changes to support Alpaca containers would also support on-prem installations.

See #152 to run the tests via the web client url (currently still in the pre-release version of AL Test Runner).

**Comment by bjarkihall** [comment]
Thanks James, sorry I didn't see this until now, I'll try testing it out tomorrow or early next week.
